src/compiler/checker.ts

```diff
@@ -439,7 +439,33 @@ module ts {
                     var declaration = forEach(result.declarations, d => isBlockOrCatchScoped(d) ? d : undefined);
 
                     Debug.assert(declaration !== undefined, "Block-scoped variable declaration is undefined");
-                    if (!isDefinedBefore(declaration, errorLocation)) {

```

Comment:
Can you move everything guarded by `if (nameNotFoundMessage)` into another function?